File: app/src/main/java/com/pietrantuono/podcasts/subscribedpodcasts/model/SubscribedPodcastModelImpl.java
package com.pietrantuono.podcasts.subscribedpodcasts.model;


import com.pietrantuono.podcasts.addpodcast.model.pojos.Podcast;
import com.pietrantuono.podcasts.addpodcast.repository.repository.Repository;

import java.util.List;

import rx.Observable;
import rx.Observer;
import rx.Subscription;

public class SubscribedPodcastModelImpl implements SubscribedPodcastModel {
    private final Repository repository;
    private Subscription subscription;

    public SubscribedPodcastModelImpl(Repository repository) {
        this.repository = repository;
    }

    @Override
    public void subscribeToSubscribedPodcasts(Observer<List<Podcast>> observer) {
        Observable<List<Podcast>> observable = repository.getSubscribedPodcasts(observer);
        subscription = observable.subscribe(observer);
    }
}

File: app/src/main/java/com/pietrantuono/podcasts/subscribedpodcasts/model/SubscribedPodcastModelImpl.kt
package com.pietrantuono.podcasts.subscribedpodcasts.model


import models.pojos.Podcast
import repo.repository.PodcastRepo
import rx.Observable
import rx.Observer
import rx.Subscription

class SubscribedPodcastModelImpl(private val repository: PodcastRepo) : SubscribedPodcastModel {
    private var subscription: Subscription? = null
    private var observable: Observable<List<Podcast>>? = null

    override fun subscribeToSubscribedPodcasts(observer: Observer<List<Podcast>>) {
        if (observable == null) {
            observable = repository.getSubscribedPodcasts()
        }
        subscription = observable?.subscribe(observer)
    }

    override fun unsubscribe() {
        subscription?.unsubscribe()
    }
}
